configs/kempner_institute/7b_Olmo_tatm.yaml

```diff
@@ -0,0 +1,168 @@
+  block_type: sequential
+  layer_norm_type: default
+  layer_norm_with_affine: true
+  bias_for_layer_norm: false
+  attention_layer_norm_with_affine: true
+  activation_type: gelu
+  residual_dropout: 0.0
+  embedding_dropout: 0.0
+  max_sequence_length: 2048
+  vocab_size: 32100           # depends on the tokenizer
+  embedding_size: 32128 
+  eos_token_id: 1
+  pad_token_id: 0
+  init_device: cuda
+  init_fn: mitchell
+
+compile: # causes instability on AMD GPUs
+  fullgraph: false
+
+activation_checkpointing: whole_layer
+
+optimizer:
+  name: adamw
+  learning_rate: 3.0e-4
+  weight_decay: 0.1
+  betas:
+  - 0.9
+  - 0.95
+  metrics_log_interval: 100
+
+scheduler:
+  name: cosine_with_warmup
+  t_warmup: 1000
+  alpha_f: 0.1
+  grad_clip_warmup_steps: 1000
+  grad_clip_warmup_factor: 10.0
+
+tokenizer:
+  identifier: t5-base
+  truncate_direction: right
+
+save_folder: ${oc.env:CHECKPOINTS_PATH}
+remote_save_folder: null
+save_overwrite: false
+# Sharded checkpoints (best for restarts)
+save_interval: 100
+save_num_checkpoints_to_keep: -1
+# Unsharded checkpoints (for final storage)
+save_interval_unsharded: null
+save_num_unsharded_checkpoints_to_keep: -1
+
+load_path: null
+
+max_duration: 515  # 135M tokens (global_train_batch_size * max_sequence_length * max_duration)
+global_train_batch_size: 128 # 4 GPUs * 32
+device_train_microbatch_size: 32
+time_limit: null
+
+precision: amp_bf16
+
+ddp:
+  grad_sync_mode: batch
+
+distributed_strategy: fsdp
+
+fsdp:
+  wrapping_strategy: by_block
+  precision: mixed
+  sharding_strategy: FULL_SHARD
+
+max_grad_norm: 1.0
+max_grad_norm_ratio: null
+
+speed_monitor:
+  window_size: 20
+
+eval_interval: ${save_interval}
+eval_subset_num_batches: 5 # limit how many batches to evaluate on (set to -1 for all batches)
+device_eval_batch_size: ${device_train_microbatch_size}
+
+evaluators:
+  - label: all-small-ppl-validation
+    data:
+      datasets:
+        c4_val: ${path.glob:/n/kempner_shared/Everyone/data/dolma/preprocessed/t5-base/c4_val/*.npy}
+      drop_last: true
+
+  ##########################
+  # Downstream evaluations #
+  ##########################
+  - label: piqa
+    type: downstream
+
+  - label: hellaswag
+    type: downstream
+
+  - label: winogrande
+    type: downstream
+
+  - label: openbook_qa
+    type: downstream
+
+  # - label: boolq  # requires implemention of the pmi_dc matrix
+    # type: downstream
+
+  - label: sciq
+    type: downstream
+
+  - label: arc_easy
+    type: downstream
+
+  # - label: arc_challenge  # requires implemention of the pmi_dc matrix
+  #   type: downstream
+
+  - label: copa
+    type: downstream
+
+  - label: rte
+    type: downstream
+
+  - label: commitment_bank
+    type: downstream
+
+  - label: mrpc
+    type: downstream
+
+  - label: sst2
+    type: downstream
+
+data:
+  pad_direction: right
+  num_workers: 2
+  drop_last: true
+  pin_memory: true
+  prefetch_factor: 1
+  persistent_workers: true
+  timeout: 0
+  custom_dataset:
```

Comment:
Just to make sure, are these configs under ‘custom_dataset’ general and not specific to ‘tatm’?
For example maybe adding a sample configuration for another well-known data loader/library that researcher are more familiar with in documents and making ng sure it is working with that config as well.

Reply:
Should be nothing `tatm` specific here since `tatm` datasets are just implementations of the Pytorch Dataset interface. The only `tatm` specific change I made was adding code to handle datasets that return dataclass objects as well as dictionaries or raw tensors, which also isn't dependent on the structure of `tatm`. I can test it out with mosaic-ml's streaming library to see if we're able to get similar results there

Reply:
Had to make one additional change to support `UserDict` objects (direct outputs of the AutoTokenizer object that `streaming` uses for their c4 implementation, but confirming that this works with MosaicML streaming datasets